### PR TITLE
feat(behavior_path_planner): extend geometric parallel parking feature to both sides

### PR DESCRIPTION
**Pull request by TakumIto**
## Description

Extend geometric pull over/out feature to both sides.
Main modification is changes in geometric_parallel_parking.cpp.

## Tests performed

All pairs of left/right side and pull over/out feature works.

https://evaluation.tier4.jp/evaluation/reports/c0776ac5-0fa7-5cd0-9cb9-6f98e28f8bfb?project_id=prd_jt
https://evaluation.tier4.jp/evaluation/reports/98969c9e-1344-543f-81e2-8755094657a1?project_id=prd_jt

## Interface changes

none
## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kyoichi-sugahara** [comment]
@TakumIto 
Suteki!!!!

File: planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp [hunk 1]
```diff
@@ -407,48 +414,64 @@ std::vector<PathWithLaneId> GeometricParallelParking::planOneTrial(
...
+      is_forward, is_forward);
+    path_turn_second.header = route_handler->getRouteHeader();
+  }
 
   // Need to add straight path to last right_turning for parking in parallel
   if (std::abs(end_pose_offset) > 0) {
     PathPointWithLaneId straight_point{};
     straight_point.point.pose = goal_pose;
-    path_turn_right.points.push_back(straight_point);
+    // setLaneIds(straight_point);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp:473]:
```suggestion
```
not needed?

File: planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp [hunk 2]
```diff
@@ -377,18 +381,21 @@ std::vector<PathWithLaneId> GeometricParallelParking::planOneTrial(
   const double goal_yaw = tf2::getYaw(arc_end_pose.orientation);
   const double psi = normalizeRadian(self_yaw - goal_yaw);
 
-  const Pose Cr = calcOffsetPose(arc_end_pose, 0, -R_E_r, 0);
-  const double d_Cr_Einit = calcDistance2d(Cr, start_pose);
+  const Pose Cfar = left_side_parking ? calcOffsetPose(arc_end_pose, 0, -R_E_far, 0)
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp:384]: C_far seems better same to R_E_far

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp]: image

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp]: 2e45e98

File: planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp [hunk 3]
```diff
@@ -407,48 +414,63 @@ std::vector<PathWithLaneId> GeometricParallelParking::planOneTrial(
...
+    path_turn_first = generateArcPath(
+      C_near, R_E_near, M_PI_2, normalizeRadian(M_PI_2 + theta_near), arc_path_interval,
+      !is_forward, is_forward);
+    path_turn_first.header = route_handler->getRouteHeader();
+
+    path_turn_second = generateArcPath(
+      C_far, R_E_far, normalizeRadian(psi - M_PI_2 + theta_near), -M_PI_2, arc_path_interval,
+      is_forward, is_forward);
+    path_turn_second.header = route_handler->getRouteHeader();
+  }
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp]: fixed in [reafactor](https://github.com/autowarefoundation/autoware.universe/pull/4881/commits/b67321149d298ec8046f5160cc9b4ba22c6886a0)

**Comment by kosuke55** [comment]
great!!!!

https://github.com/autowarefoundation/autoware.universe/assets/39142679/d81342c2-5db0-4ba2-81ca-5d842b8824e3



https://github.com/autowarefoundation/autoware.universe/assets/39142679/9e0f1de4-157b-4916-849a-19e850f5cd4c




https://github.com/autowarefoundation/autoware.universe/assets/39142679/205a42aa-3b1b-4ce9-9544-c306894908bc



https://github.com/autowarefoundation/autoware.universe/assets/39142679/960fd654-b582-4bd4-bd1a-7dbac1be1781

File: planning/behavior_path_planner/include/behavior_path_planner/utils/geometric_parallel_parking/geometric_parallel_parking.hpp [hunk 4]
```diff
@@ -121,10 +122,10 @@ class GeometricParallelParking
 
   void clearPaths();
   std::vector<PathWithLaneId> planOneTrial(
-    const Pose & start_pose, const Pose & goal_pose, const double R_E_r,
+    const Pose & start_pose, const Pose & goal_pose, const double R_E_far,
     const lanelet::ConstLanelets & road_lanes, const lanelet::ConstLanelets & shoulder_lanes,
-    bool is_forward, const double end_pose_offset, const double lane_departure_margin,
-    const double arc_path_interval);
+    bool is_forward, bool left_side_parking, const double end_pose_offset,
```
Comment (kosuke55) [inline planning/behavior_path_planner/include/behavior_path_planner/utils/geometric_parallel_parking/geometric_parallel_parking.hpp:127]: const

Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/utils/geometric_parallel_parking/geometric_parallel_parking.hpp]: fixed in [const](https://github.com/autowarefoundation/autoware.universe/pull/4881/commits/9ac9482626b492e1ca1b5110c316f80282e92ed6)

File: planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp [hunk 5]
```diff
@@ -502,8 +526,8 @@ std::vector<PathWithLaneId> GeometricParallelParking::planOneTrial(
   arc_end_pose_ = arc_end_pose;
 
   // debug
-  Cr_ = Cr;
-  Cl_ = Cl;
+  Cr_ = left_side_parking ? C_far : C_near;
+  Cl_ = left_side_parking ? C_near : C_far;
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/geometric_parallel_parking/geometric_parallel_parking.cpp:530]: (I did not check this debug marker)

**Review by kosuke55 — APPROVED** [review]
LGTM!!! thanks!!

**Pull request merged**